### PR TITLE
Fix potential corrupt state in AbstractPluginManager

### PR DESCRIPTION
**Pull request by slovdahl**
Related to #459.

### CODE REVIEW

File: pf4j/src/test/java/org/pf4j/DefaultPluginManagerTest.java [hunk 1]
```diff
@@ -190,4 +190,44 @@ public void deleteJarPlugin() throws Exception {
...
+
+        PluginZip pluginZip2 = new PluginZip.Builder(pluginsPath.resolve("my-second-plugin-2.2.2.zip"), "myPlugin2")
+            .pluginVersion("2.2.2")
+            .build();
+
+        pluginManager.loadPlugins();
+        pluginManager.startPlugins();
+
+        // myPlugin2 should have been started at this point.
+        assertEquals(PluginState.STARTED, pluginManager.getPlugin(pluginZip2.pluginId()).getPluginState());
```
Comment (slovdahl) [inline pf4j/src/test/java/org/pf4j/DefaultPluginManagerTest.java:209]: Right now, the test will fail at this point.

Comment (slovdahl) [reply pf4j/src/test/java/org/pf4j/DefaultPluginManagerTest.java:209]: Some suggested fixes pushed in https://github.com/pf4j/pf4j/pull/558/commits/b96bb93b8ceb44a43c6bb95df3057bbd59c471f6.

**Comment by decebals** [comment]
I test the test supplied by you in this PR, it passes but the test introduced in #460 failed at https://github.com/pf4j/pf4j/pull/460/files#diff-6dbe4409bbc76e0eb975d634fb3aa8deac4debe8e9cdbb872812861f5b304056R206

**Comment by decebals** [comment]
From what I understand this part with `unloadPlugin` is sensible because the plugin is removed from list with plugins.
I think that it's fine to have all plugins in list but with different states.

**Comment by decebals** [comment]
The current `AbstractPluginManager#resolvePlugins`  comes with an ALL OR NOTHING implementation . According [code](https://github.com/pf4j/pf4j/blob/master/pf4j/src/main/java/org/pf4j/AbstractPluginManager.java#L791) if they plugins with missing dependencies the entire resolve plugins process fails (an exception is thrown) . I must admit that in this scenario to suppress the exception in `loadPlugins` (only an error message in log) is not perfect because the application does not have a feedback that the process of loading plugins went as it should.

**Comment by decebals** [comment]
On the other side to go only with the plugins that could be solved, I don't think it's a fit all solution. Maybe we need to implement a strategy pattern here and the application to choose one.

**Comment by slovdahl** [comment]
> I test the test supplied by you in this PR, it passes but the test introduced in https://github.com/pf4j/pf4j/pull/460 failed at https://github.com/pf4j/pf4j/pull/460/files#diff-6dbe4409bbc76e0eb975d634fb3aa8deac4debe8e9cdbb872812861f5b304056R206

Ah, yes, of course :smile: Because this change avoids the problem in the first place (that we need to manually unload a plugin that was partially loaded).

**Comment by slovdahl** [comment]
> From what I understand this part with unloadPlugin is sensible because the plugin is removed from the list with plugins.
I think that it's fine to have all plugins in list but with different states.

So you think the two changes in `AbstractPluginManager#unloadPlugin` could be merged as is? In that case I can extract them to a separate PR, and see if I can write some specific tests for them.

**Comment by slovdahl** [comment]
> I must admit that in this scenario to suppress the exception in loadPlugins (only an error message in log) is not perfect because the application does not have a feedback that the process of loading plugins went as it should.

:+1: 

> On the other side to go only with the plugins that could be solved, I don't think it's a fit all solution. Maybe we need to implement a strategy pattern here and the application to choose one.

Yep, agreed, this might not be for everyone. In our context, it's what we want, but some kind of strategy sounds like an even better solution. Do you have any concrete ideas in mind?

**Comment by decebals** [comment]
@slovdahl 
> Do you have any concrete ideas in mind?

Yes, see https://github.com/pf4j/pf4j/pull/564.

**Comment by decebals** [comment]
Replaced by https://github.com/pf4j/pf4j/pull/564.

**Pull request closed without merging**